### ISSUE TITLE
False Positive: Null-initialization idioms and `try-with-resources` cases are misreported as dead stores.

### ISSUE THREAD

**Issue by Carlson-JLQ** [pmd/pmd#6541]
# False Positive: Null-initialization idioms and `try-with-resources` cases are misreported as dead stores.

Version
7.22.0

## Checker
- Checker id: `category/java/bestpractices.xml/UnusedAssignment`
- Checker description: This rule detects assignments to variables where the assigned value is never read before being overwritten or going out of scope.

## Description of the false positive
`diff/output` marks 3 false positives for `category/java/bestpractices.xml/UnusedAssignment`. Across these samples, the checker tends to fire on the outer syntax while missing code-level cues that make the pattern harmless, exempted, or at least much weaker than a real bug. The recurring themes are resource-management structure.

### `NegCase1.java`
Source: `output/codeql/DeadStoreOfLocal/src/main/java/scensct/core/neg/NegCase1.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. A human reader would likely classify this as an allowed or intentionally structured pattern rather than a bug. That matches the benchmark's stated intent: A local variable declared with a null initializer that is overwritten before being read should not be flagged as a dead store.

```java
// A local variable declared with a null initializer that is overwritten before being read should not be flagged as a dead store.
package scensct.core.neg;

public class NegCase1 {
    public static void main(String[] args) {
        // Scenario: Reference type with null initializer, overwritten before read
        String s = null; // Initializer excluded from reporting due to type and value
        s = "Hello";
        System.out.println(s);
    }
}
```

### `NegCase2.java`
Source: `output/codeql/DeadStoreOfLocal/src/main/java/scensct/core/neg/NegCase2.java`

The sample keeps the benchmark's target pattern but expresses it in a slightly non-canonical source form. A human reader would likely classify this as an allowed or intentionally structured pattern rather than a bug. That matches the benchmark's stated intent: A dead store assignment that is a common idiom (e.g., int x = 0) should not be flagged as a dead store.

```java
// A dead store assignment that is a common idiom (e.g., int x = 0) should not be flagged as a dead store.
package scensct.core.neg;

public class NegCase2 {
    public static void main(String[] args) {
        // Scenario: Common idiom initializer, overwritten before read but excluded
        int x = 0; // Common idiom, excluded by checker rules
        x = 5;
        System.out.println(x);
    }
}
```

### `NegCase2.java`
Source: `output/sonarqube/DeadStoreCheck/src/main/java/scensct/core/neg/NegCase2.java`

The assignment that matters appears inside a `try-with-resources` block, which often changes the local control-flow shape seen by the checker. What looks like a dead store is part of a benchmark carveout around resource-management structure rather than an obviously useless assignment. That matches the benchmark's stated intent: A method with a try-with-resources statement containing a dead store should not be flagged as dead store.

```java
// A method with a try-with-resources statement containing a dead store should not be flagged as dead store.
package scensct.core.neg;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

public class NegCase2 {
    public void method() {
        try (BufferedReader br = new BufferedReader(new FileReader("test.txt"))) {
            String line = br.readLine();
            int unused = 42; // Dead store assignment inside try-with-resources
            unused = 43; // Reassignment before any read
            System.out.println(line);
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}
```

## Cause analysis
The rule is documented as follows: This rule detects assignments to variables where the assigned value is never read before being overwritten or going out of scope.

In this set the problem shows up around resource-management structure. The implementation appears to lean too heavily on surface indicators and not enough on what the code actually does.

That is why these reports read like over-approximation. The code often retains a suspicious signature or control-flow skeleton, but once you read the body, the benchmark's claimed exemption is visible in the source itself.

For `category/java/bestpractices.xml/UnusedAssignment`, the practical improvement would be to model those source-level exemptions more explicitly instead of reporting every syntactic near-match.

## References
None known.
